sdk/scheduler/src/main/java/com/mesosphere/sdk/scheduler/decommission/DecommissionPlanFactory.java

```diff
@@ -0,0 +1,289 @@
+                // (pod2 before pod1, and unknown before pod2)
+                return podTypeIndex - other.podTypeIndex;
+            } else if (podTypeIndex == -1 && !podType.equals(other.podType)) {
+                // The pod types are both unlisted, but are not equal to each other. Just order them alphabetically.
+                // (unknownA before unknownB)
+                return podType.compareTo(other.podType);
+            }
+            // The pod types are indentical. Order them according to their index. (pod-3 before pod-2)
+            return other.podIndex - podIndex;
+        }
+
+        public String getPodName() {
+            return PodInstance.getName(podType, podIndex);
+        }
+
+        @Override
+        public String toString() {
+            return getPodName();
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            return EqualsBuilder.reflectionEquals(this, o);
+        }
+
+        @Override
+        public int hashCode() {
+            return HashCodeBuilder.reflectionHashCode(this);
+        }
+    }
+
+    /**
+     * Returns a mapping of pods to be decommissioned with affected tasks within those pods. The returned mapping will
+     * be in the order that the pods should be decommissioned.
+     */
+    @VisibleForTesting
+    static SortedMap<PodKey, Collection<Protos.TaskInfo>> getPodsToDecommission(
+            ServiceSpec serviceSpec, Collection<Protos.TaskInfo> tasks) {
+        // If multiple pod types are being decommissioned, they should be decommissioned in the reverse of the order
+        // that they're declared in the ServiceSpec (opposite direction of default deployment)
+        List<String> orderedPodTypes =
+                serviceSpec.getPods().stream().map(PodSpec::getType).collect(Collectors.toList());
+        Collections.reverse(orderedPodTypes);
+
+        Map<String, Integer> expectedPodCounts =
+                serviceSpec.getPods().stream().collect(Collectors.toMap(PodSpec::getType, PodSpec::getCount));
+        LOGGER.info("Expected pod counts: {}", expectedPodCounts);
+        SortedMap<PodKey, Collection<Protos.TaskInfo>> podsToDecommission = new TreeMap<>();
+        for (Protos.TaskInfo task : tasks) {
+            final PodKey podKey;
+            try {
+                TaskLabelReader labelReader = new TaskLabelReader(task);
+                podKey = new PodKey(labelReader.getType(), labelReader.getIndex(), orderedPodTypes);
+            } catch (TaskException e) {
+                LOGGER.error(String.format(
+                        "Failed to retrieve task metadata. Omitting task from decommission: %s", task.getName()), e);
+                continue;
+            }
+
+            Integer expectedPodCount = expectedPodCounts.get(podKey.podType);
+            if (expectedPodCount == null) {
+                LOGGER.info("Scheduling '{}' for decommission: '{}' is not present in service spec: {}",
```

Comment:
This seems like very dangerous behavior. It could mean dropping an entire pod type on accident through a bad update.

What is the current behavior if an entire unknown pod type is found in the state store?

Reply:
If we do want this behavior, we should make it clear that any update dropping a pod type is irreversible.

Reply:
We definitely want the ability to drop pod-types.  Very good for schedulers which want to generate/remove pod types programmatically.

Reply:
I'm fine with clear documentation in the dev guide that old pod types will be garbage collected. Rather, that they create one-way upgrades if you do it in a package upgrade.

I hadn't thought of the generate/remove pod types use case. Neat.

Reply:
Updated #1858